# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
# Import your libraries:

%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sqlalchemy import create_engine


# Challenge 0 - Load,Query and Create connection of your Dataset

#### A lot of the times you won't have files already saved in an Excel or CSV file for you to prepare your data, implying that a lot of times you'll be extracting data from a Data Warehouse or Data Lake, majority of times through SQL.
#### A couple of times you may want to do some queries on your table in mySQL to get a slight view on what you have in hands,
#### so let's simulate that!

#### First we'll need to create a database and table in mySQL:

##### 1º- Open the austin_weather.sql file in MySQL Workbench and run the script into a desired schema.

#### 2º- As we are in mySQL Workbench, we can do some queries there, to get an overview on some characteristics of our data:
 - a) How many days are recorded in the dataset?
 - b) What is the day with the Highest Temperature in Fahrenheit (column TempHighF)
 - c) What is the average Humidity across all days? (column HumidityAvgPercent)
 - d) Top 10 days, where SeaLevelPressureAvgInches is the highest, knowing DewPointAvgF is higher than 28 ?

In [ ]:
# Write your answers below:

# a) 1319
# b)2013-12-21
# c)66.57088
# d)30.54

#### Now that you explored a couple of elements in your table, let's bring your table into this jupyter notebook, by creating a Python-SQL connection like you did on MySQL Project!
#### In case you need a little refresher check this [link](https://www.dataquest.io/blog/sql-insert-tutorial/).
##### 1º - Create a connection using sqlalchemy from python to mysql 
##### 2º- Load the table into a variable called weather_df


In [23]:
# Your code here
from sqlalchemy import create_engine
import pandas as pd

user = 'root'
password = 'Pastillas2#'
host = '127.0.0.1'  # o la IP de tu servidor
port = '3306'     # puerto por defecto de MySQL
database = 'weather'

#CREATE THE CONNECTION 
connection_str = f"mysql+pymysql://{user}:{password}@{host}:{port}/{database}"

engine = create_engine(connection_str)

# Load the data base
weather_df = pd.read_sql('SELECT * FROM austin_weather', con=engine)
weather_df.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]


# Challenge 1 - Describe the Dataset

#### Describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [24]:
# Your code here
weather_df.dtypes

Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                         object
DewPointAvgF                          object
DewPointLowF                          object
HumidityHighPercent                   object
HumidityAvgPercent                    object
HumidityLowPercent                    object
SeaLevelPressureHighInches            object
SeaLevelPressureAvgInches             object
SeaLevelPressureLowInches             object
VisibilityHighMiles                   object
VisibilityAvgMiles                    object
VisibilityLowMiles                    object
WindHighMPH                           object
WindAvgMPH                            object
WindGustMPH                           object
PrecipitationSumInches                object
Events                                object
dtype: object

In [25]:
# Your code here
weather_df.describe()

Date    TempHighF     TempAvgF     TempLowF
count                 2638  2638.000000  2638.000000  2638.000000
mean   2015-10-11 00:00:00    80.862775    70.642911    59.902957
min    2013-12-21 00:00:00    32.000000    29.000000    19.000000
25%    2014-11-15 06:00:00    72.000000    62.000000    49.000000
50%    2015-10-11 00:00:00    83.000000    73.000000    63.000000
75%    2016-09-04 18:00:00    92.000000    83.000000    73.000000
max    2017-07-31 00:00:00   107.000000    93.000000    81.000000
std                    NaN    14.763723    14.043240    14.187957

In [26]:
# Your code here
weather_df.head(5)

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [28]:
# Your code here
weather_df['DewPointAvgF'].unique()


array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [31]:
# Your observation here
#There is a row that it does not have a number, instead it has '-'
#For that reason is an 'Object' and not 'int64'

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [34]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [38]:
# Your code here

# Convertir todas esas columnas a tipo numérico
# Si hay errores (por ejemplo, strings como 'error' o 'NaN'), se convierten en valores faltantes (NaN)
weather_df[wrong_type_columns] = weather_df[wrong_type_columns].apply(
    pd.to_numeric, errors='coerce'
)

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [39]:
# Your code here
weather_df.dtypes

Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                        float64
DewPointAvgF                         float64
DewPointLowF                         float64
HumidityHighPercent                  float64
HumidityAvgPercent                   float64
HumidityLowPercent                   float64
SeaLevelPressureHighInches           float64
SeaLevelPressureAvgInches            float64
SeaLevelPressureLowInches            float64
VisibilityHighMiles                  float64
VisibilityAvgMiles                   float64
VisibilityLowMiles                   float64
WindHighMPH                          float64
WindAvgMPH                           float64
WindGustMPH                          float64
PrecipitationSumInches               float64
Events                                object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [43]:
# Your code here
weather_df.isnull().any(axis=1)


0       False
1       False
2       False
3       False
4        True
        ...  
2633    False
2634    False
2635    False
2636    False
2637    False
Length: 2638, dtype: bool

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [48]:
# Your code here

total_rows = len(weather_df)
missing_values = weather_df.isnull().sum()

print("Number of rows:", total_rows)
print("Rows with empty values:", missing_values)

Number of rows: 2638
Rows with empty values: Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                  14
DewPointAvgF                   14
DewPointLowF                   14
HumidityHighPercent             4
HumidityAvgPercent              4
HumidityLowPercent              4
SeaLevelPressureHighInches      6
SeaLevelPressureAvgInches       6
SeaLevelPressureLowInches       6
VisibilityHighMiles            24
VisibilityAvgMiles             24
VisibilityLowMiles             24
WindHighMPH                     4
WindAvgMPH                      4
WindGustMPH                     8
PrecipitationSumInches        248
Events                          0
dtype: int64


#### Calculate the ratio of missing rows to total rows

In [52]:
# Your code here
missing_ratio = missing_values / total_rows
print(missing_ratio)

Date                          0.000000
TempHighF                     0.000000
TempAvgF                      0.000000
TempLowF                      0.000000
DewPointHighF                 0.005307
DewPointAvgF                  0.005307
DewPointLowF                  0.005307
HumidityHighPercent           0.001516
HumidityAvgPercent            0.001516
HumidityLowPercent            0.001516
SeaLevelPressureHighInches    0.002274
SeaLevelPressureAvgInches     0.002274
SeaLevelPressureLowInches     0.002274
VisibilityHighMiles           0.009098
VisibilityAvgMiles            0.009098
VisibilityLowMiles            0.009098
WindHighMPH                   0.001516
WindAvgMPH                    0.001516
WindGustMPH                   0.003033
PrecipitationSumInches        0.094011
Events                        0.000000
dtype: float64


As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [54]:
# Your code here
weather_df.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                  14
DewPointAvgF                   14
DewPointLowF                   14
HumidityHighPercent             4
HumidityAvgPercent              4
HumidityLowPercent              4
SeaLevelPressureHighInches      6
SeaLevelPressureAvgInches       6
SeaLevelPressureLowInches       6
VisibilityHighMiles            24
VisibilityAvgMiles             24
VisibilityLowMiles             24
WindHighMPH                     4
WindAvgMPH                      4
WindGustMPH                     8
PrecipitationSumInches        248
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [ ]:
# Your code here
#9.4%

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [57]:
# Your code here 
#axis=1 to drop a column instead of a row 
#inplace= true to actually modify the original dataset instead of just a copy
weather_df.drop('PrecipitationSumInches', axis=1, inplace=True)

# Print `austin` to confirm the column is indeed removed

print(weather_df.columns)

KeyError: "['PrecipitationSumInches'] not found in axis"

In [58]:

print(weather_df.columns)

Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'Events'],
      dtype='object')


#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [59]:
# Your code here
weather_fixed = weather_df.interpolate(method='linear', inplace=False)

/var/folders/tc/dvbccp5x0rzg3_4x3qdj46zh0000gn/T/ipykernel_22726/3979662961.py:2: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  weather_fixed = weather_df.interpolate(method='linear', inplace=False)


#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [65]:
# Your code here
weather_fixed.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [66]:
# Your code here:
weather_fixed['Events'].value_counts()

#Hace un conteo de los valores de la columna. Cuantas veces ese valor aparece en la columna

Events
                             1806
Rain                          384
Rain , Thunderstorm           274
Fog , Rain , Thunderstorm      66
Fog                            42
Thunderstorm                   34
Fog , Rain                     28
Rain , Snow                     2
Fog , Thunderstorm              2
Name: count, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [67]:
# Your answer:
#Rain
#Rain, Thunderstorm


#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [68]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for event in event_list:
    weather_fixed[event] = 0

# Print your new dataframe to check whether new columns have been created:

print(weather_fixed.columns)

Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'Events', 'Snow', 'Fog', 'Rain', 'Thunderstorm'],
      dtype='object')


#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [71]:
# Your code here
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

for event in event_list:
    # Revisar si la columna 'Events' contiene el texto del evento
    weather_fixed[event] = weather_fixed['Events'].str.contains(event, na=False).astype(int)

#austin_fixed['Events'].str.contains(event, na=False)
#Revisa si la palabra 'Rain', 'Fog', etc., aparece en cada fila de la columna 'Events'.
#Devuelve una serie de valores booleanos (True o False).
# .astype(int)Convierte True en 1 y False en 0.
# Finalmente, eso se asigna a la columna correspondiente en austin_fixed[event].


#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [70]:
# Your code here
weather_fixed.head(19)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   
5  2013-12-26         57        48        39           39.0          36.0   
6  2013-12-27         60        53        45           41.0          39.0   
7  2013-12-28         62        51        40           43.0          39.0   
8  2013-12-29         64        50        36           49.0          41.0   
9  2013-12-30         44        40        35           31.0          26.0   
10 2013-12-31         55        46        36           31.0          28.0   
11 2014-01-01         69        54        39           51.0          42.0   
12 2014-01-02         55        44        33           39.0          26.0   
13 2014-01-03         58        43        28           37.0          22.0   
14 2014-01-04         71        57        42           55.0          48.0   
15 2014-01-05         59        47        34           54.0          32.0   
16 2014-01-06         36        29        22           15.0           8.0   
17 2014-01-07         48        35        22           29.0          11.0   
18 2014-01-08         53        47        40           51.0          45.0   

    DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0           43.0                 93.0                75.0                57.0   
1           28.0                 93.0                68.0                43.0   
2           23.0                 76.0                52.0                27.0   
3           21.0                 89.0                56.0                22.0   
4           36.0                 86.0                71.0                56.0   
5           33.0                 79.0                63.0                47.0   
6           37.0                 83.0                65.0                47.0   
7           33.0                 92.0                64.0                36.0   
8           28.0                 92.0                76.0                60.0   
9           21.0                 75.0                60.0                45.0   
10          23.0                 76.0                54.0                32.0   
11          30.0                 83.0                68.0                52.0   
12          19.0                 83.0                55.0                26.0   
13          18.0                 75.0                49.0                22.0   
14          38.0                 89.0                68.0                47.0   
15          15.0                 87.0                59.0                31.0   
16           2.0                 50.0                38.0                26.0   
17           4.0                 68.0                43.0                17.0   
18          30.0                 93.0                75.0                57.0   

    ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0   ...                 7.0                 2.0         20.0         4.0   
1   ...                10.0                 5.0         16.0         6.0   
2   ...                10.0                10.0          8.0         3.0   
3   ...                10.0                 7.0         12.0         4.0   
4   ...                10.0                 7.0         10.0         2.0   
5   ...                 9.0                 7.0         12.0         3.0   
6   ...                 9.0                 7.0          7.0         1.0   
7   ...                10.0                 7.0         10.0         2.0   
8   ...                10.0                 4.0         17.0         5.0   
9   ...                10.0                10.0         13.0         5.0   
10  ...    

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [73]:
# Your code here
weather_fixed.drop('Events', axis=1, inplace=True)


KeyError: "['Events'] not found in axis"

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [75]:
# Your code here
import datetime

# Función para convertir Timestamp a ordinal
def convert_to_ordinal(dt_obj):
    return dt_obj.toordinal()

# Aplicar la función
weather_fixed['Date'] = weather_fixed['Date'].apply(convert_to_ordinal)



#### Print `austin_fixed` to check your `Date` column.

In [76]:
weather_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0  ...                 10.0                 7.0                 2.0   
1  ...                 10.0                10.0                 5.0   
2  ...                 10.0                10.0                10.0   
3  ...                 10.0                10.0                 7.0   
4  ...                 10.0                10.0                 7.0   

   WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0         20.0         4.0         31.0     0    0     1             1  
1         16.0         6.0         25.0     0    0     0             0  
2          8.0         3.0         12.0     0    0     0             0  
3         12.0         4.0         20.0     0    0     0             0  
4         10.0         2.0         16.0     0    0     0             0  

[5 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [82]:
# Your code here:
#Create the response variable (target)
y = weather_fixed['TempAvgF']

#Create the predictor variables
X_columns = [
    'Date',
    'TempHighF',
    'TempLowF',
    'DewPointHighF',
    'DewPointAvgF',
    'DewPointLowF',
    'HumidityHighPercent',
    'HumidityAvgPercent',
    'HumidityLowPercent',
    'SeaLevelPressureHighInches',
    'SeaLevelPressureAvgInches',
    'SeaLevelPressureLowInches',
    'VisibilityHighMiles',
    'VisibilityAvgMiles',
    'VisibilityLowMiles',
    'WindHighMPH',
    'WindAvgMPH',
    'WindGustMPH',
    'Snow',
    'Fog',
    'Rain',
    'Thunderstorm'
]

X = weather_fixed[X_columns]

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [83]:
#Your code here:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [87]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,                  # your predictors and target
    test_size=0.2,         # 20% for testing
    random_state=42        # ensures reproducibility
)


#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:
